datafusion-examples/examples/memory_pool_execution_plan.rs

```diff
@@ -0,0 +1,324 @@
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! This example demonstrates how to implement custom ExecutionPlans that properly
+//! use memory tracking through TrackConsumersPool.
+//!
+//! This shows the pattern for implementing memory-aware operators that:
+//! - Register memory consumers with the pool
+//! - Reserve memory before allocating
+//! - Handle memory pressure by spilling to disk
+//! - Release memory when done
+
+use arrow::array::{Int32Array, StringArray};
+use arrow::record_batch::RecordBatch;
+use arrow_schema::{DataType, Field, Schema, SchemaRef};
+use datafusion::datasource::{memory::MemTable, DefaultTableSource};
+use datafusion::error::{DataFusionError, Result};
+use datafusion::execution::memory_pool::{MemoryConsumer, MemoryReservation};
+use datafusion::execution::runtime_env::RuntimeEnvBuilder;
+use datafusion::execution::{SendableRecordBatchStream, TaskContext};
+use datafusion::logical_expr::LogicalPlanBuilder;
+use datafusion::physical_plan::stream::RecordBatchStreamAdapter;
+use datafusion::physical_plan::{
+    DisplayAs, DisplayFormatType, ExecutionPlan, PlanProperties, Statistics,
+};
+use datafusion::prelude::*;
+use futures::stream::{StreamExt, TryStreamExt};
+use std::any::Any;
+use std::fmt;
+use std::sync::Arc;
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    println!("=== DataFusion ExecutionPlan Memory Tracking Example ===\n");
+
+    // Set up a runtime with memory tracking
+    // Set a low memory limit to trigger spilling on the second batch
+    let runtime = RuntimeEnvBuilder::new()
+        .with_memory_limit(15000, 1.0) // Allow only enough for 1 batch at once
+        .build_arc()?;
+
+    let config = SessionConfig::new().with_coalesce_batches(false);
+    let ctx = SessionContext::new_with_config_rt(config, runtime.clone());
+
+    // Create some test data
+    let schema = Arc::new(Schema::new(vec![
+        Field::new("id", DataType::Int32, false),
+        Field::new("name", DataType::Utf8, false),
+    ]));
+
+    // Create smaller batches to ensure we get multiple RecordBatches from the scan
+    // Make each batch smaller than the memory limit to force multiple batches
+    let batch1 = RecordBatch::try_new(
```

Comment:
maybe this would be a good use for the https://docs.rs/arrow/latest/arrow/array/macro.record_batch.html macro